device/transport/amqp/test/_amqp_test.js

```diff
@@ -4,246 +4,1345 @@
     };
+
+    transport = new Amqp(configWithSAS, fakeBaseClient);
   });
 
   afterEach(function () {
     transport = null;
     receiver = null;
+    sender = null;
+    fakeBaseClient = null;
   });
 
-  describe('#sendMethodResponse', function() {
-    /*Tests_SRS_NODE_DEVICE_AMQP_16_019: [The `sendMethodResponse` shall throw a `ReferenceError` if the `methodResponse` object is falsy.]*/
-    [null, undefined].forEach(function(badResponse) {
-      it('throws a ReferenceError if the methodResponse object is \'' + badResponse + '\'', function() {
-        assert.throws(function() {
-          transport.sendMethodResponse(badResponse, function() {});
-        }, ReferenceError);
+  describe('Direct Methods', function () {
+    describe('#sendMethodResponse', function () {
+      /*Tests_SRS_NODE_DEVICE_AMQP_16_019: [The `sendMethodResponse` shall throw a `ReferenceError` if the `methodResponse` object is falsy.]*/
+      [null, undefined].forEach(function (badResponse) {
+        it('throws a ReferenceError if the methodResponse object is \'' + badResponse + '\'', function () {
+          assert.throws(function () {
+            transport.sendMethodResponse(badResponse, function () {});
+          }, ReferenceError);
+        });
+      });
+
+      it('calls the callback with a NotConnectedError if the device is disconnected', function (testCallback) {
+        var fakeMethodResponse = { status: 200, payload: null, requestId: uuid.v4() };
+        transport.sendMethodResponse(fakeMethodResponse, function (err) {
+          assert.instanceOf(err, errors.NotConnectedError);
+          testCallback();
+        });
+      });
+
+      it('calls the callback with a NotConnectedError if the device is connecting', function (testCallback) {
+        var fakeMethodResponse = { status: 200, payload: null, requestId: uuid.v4() };
+        fakeBaseClient.connect = sinon.stub();
+        transport.connect();
+        transport.sendMethodResponse(fakeMethodResponse, function (err) {
+          assert.instanceOf(err, errors.NotConnectedError);
+          testCallback();
+        });
+      });
+
+      /*Tests_SRS_NODE_DEVICE_AMQP_16_020: [The `sendMethodResponse` response shall call the `AmqpDeviceMethodClient.sendMethodResponse` method with the arguments that were given to it.]*/
+      it('calls the `sendMethodResponse` method on the AmqpDeviceMethodClient object', function (testCallback) {
+        var fakeMethodResponse = { status: 200, payload: null, requestId: uuid.v4() };
+        var fakeCallback = function () {};
+        sinon.spy(transport._deviceMethodClient, 'sendMethodResponse');
+
+        transport.connect(function () {
+          transport.sendMethodResponse(fakeMethodResponse, fakeCallback);
+          assert.isTrue(transport._deviceMethodClient.sendMethodResponse.calledWith(fakeMethodResponse));
+          testCallback();
+        });
+      });
+
+      it('fails if called disconnecting', function (testCallback) {
```

Comment:
>fails if called disconnecting' [](start = 10, length = 30)

So what's the rational that this fails if disconnected but that onDeviceMethod will initiate a connect.  (At least that's how the requirments read).

Reply:
sending a method response requires using the same links that were used to receive the method request. so if we're disconnecting, the response will fail.
However, any call to onDeviceMethod while disconnected is OK because we're not responding, we're establishing new links.

---
In reply to: [141121863](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141121863) [](ancestors = 141121863)